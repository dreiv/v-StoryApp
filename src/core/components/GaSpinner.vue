<script setup lang="ts">
import { computed, useCssModule, onBeforeMount } from 'vue'
import { Hourglass } from 'lucide-vue-next'
import type { FormInputProps } from '../composables/useFormInput'

export interface SpinnerProps extends FormInputProps {
  size?: 'sm' | 'default' | 'lg'
  direction?: 'horizontal' | 'vertical'
}

let reducedMotion = false
onBeforeMount(() => {
  reducedMotion = window.matchMedia('(prefers-reduced-motion: reduce)').matches
})

const {
  size = 'default',
  direction = 'horizontal',
  error,
  errorMessage,
  label,
} = defineProps<SpinnerProps>()
const style = useCssModule()

const sizeCss = computed(() => [{ [style.small]: size === 'sm', [style.large]: size === 'lg' }])
const containerCss = computed(() => [
  style.container,
  { [style.vertical]: direction === 'vertical' },
])

const aria = computed(() => ({
  'aria-label': label,
  'aria-invalid': error ? true : undefined,
  'aria-errormessage': errorMessage ?? undefined,
}))
</script>

<template>
  <div :class="containerCss" v-bind="aria">
    <slot name="icon">
      <hourglass v-if="reducedMotion" :class="[$style.icon, ...sizeCss]" />

      <svg
        v-else
        xmlns="http://www.w3.org/2000/svg"
        width="32"
        height="32"
        viewBox="0 0 32 32"
        stroke-width="6.4"
        :class="[$style.icon, ...sizeCss]"
      >
        <defs><circle id="circle" cx="16" cy="16" r="12.8" fill="none" /></defs>

        <use href="#circle" stroke="var(--ga-color-secondary)" />
        <use
          href="#circle"
          :class="$style.spinner"
          stroke="var(--ga-color-primary-70)"
          stroke-linecap="round"
          stroke-dasharray="21 84"
        />
      </svg>
    </slot>

    <span :class="[$style.label, ...sizeCss]">
      <slot>{{ label }}</slot>
    </span>
  </div>
</template>

<style module>
.container {
  display: flex;
  align-items: center;
  gap: var(--ga-size-spacing-02);

  &.vertical {
    flex-direction: column;
    gap: var(--ga-size-spacing-03);
  }
}

.icon {
  &.small {
    width: var(--ga-size-spacing-05);
    height: var(--ga-size-spacing-05);
  }
  &.large {
    width: var(--ga-size-spacing-09);
    height: var(--ga-size-spacing-09);
  }
}

@keyframes spin {
  100% {
    transform: rotate(360deg);
  }
}
.spinner {
  transform-origin: center;
  animation: spin 1.1s cubic-bezier(0.4, 0.2, 0.6, 0.8) infinite;
  will-change: transform;
}

.label {
  min-height: var(--ga-size-spacing-05);
  color: var(--ga-color-text-body);

  font-weight: var(--ga-font-weight-normal);
  font-size: var(--ga-text-sm-font-size);
  line-height: var(--ga-text-sm-line-height);
  font-family: var(--ga-font-family-primary);
  letter-spacing: var(--ga-text-sm-tracking);

  &:empty {
    display: none;
  }
}

.vertical .label {
  &.small {
    font-size: var(--ga-text-xs-font-size);
    line-height: var(--ga-text-xs-line-height);
    letter-spacing: var(--ga-text-xs-tracking);
  }

  &.large {
    font-size: var(--ga-text-md-font-size);
    line-height: var(--ga-text-md-line-height);
    letter-spacing: var(--ga-text-md-tracking);
  }
}
</style>
