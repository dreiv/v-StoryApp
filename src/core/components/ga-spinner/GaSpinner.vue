<script setup lang="ts">
import { computed, useCssModule, ref, onMounted, onUnmounted } from 'vue'
import { Hourglass } from 'lucide-vue-next'
import { type FormInputProps } from '@/core/composables/useFormInput'

export interface SpinnerProps extends FormInputProps {
  size?: 'small' | 'medium' | 'large'
  direction?: 'horizontal' | 'vertical'
}

const reducedMotion = ref(false)
let mediaQuery: MediaQueryList
let handleChange: (e: MediaQueryListEvent) => void

onMounted(() => {
  mediaQuery = window.matchMedia('(prefers-reduced-motion: reduce)')
  reducedMotion.value = mediaQuery.matches

  handleChange = (e: MediaQueryListEvent) => {
    reducedMotion.value = e.matches
  }

  mediaQuery.addEventListener('change', handleChange)
})

onUnmounted(() => {
  if (mediaQuery && handleChange) {
    mediaQuery.removeEventListener('change', handleChange)
  }
})

const {
  size = 'medium',
  direction = 'horizontal',
  error,
  errorMessage,
  label,
} = defineProps<SpinnerProps>()
const style = useCssModule()

const iconClasses = computed(() => [style.icon, style[size]])

const labelClasses = computed(() => [style.label, style[size]])

const containerClasses = computed(() => {
  const classList = [style.container]

  if (direction === 'vertical') classList.push(style.vertical)

  return classList
})

const aria = computed(() => ({
  'aria-label': label,
  'aria-invalid': error ? true : undefined,
  'aria-errormessage': errorMessage || undefined,
}))
</script>

<template>
  <div :class="containerClasses" v-bind="aria">
    <slot name="icon">
      <hourglass v-if="reducedMotion" :class="iconClasses" />

      <svg
        v-else
        xmlns="http://www.w3.org/2000/svg"
        width="32"
        height="32"
        viewBox="0 0 32 32"
        stroke-width="6.4"
        :class="iconClasses"
      >
        <defs><circle id="circle" cx="16" cy="16" r="12.8" fill="none" /></defs>

        <use href="#circle" stroke="var(--ga-color-secondary)" />
        <use
          href="#circle"
          :class="$style.spinner"
          stroke="var(--ga-color-primary-70)"
          stroke-linecap="round"
          stroke-dasharray="21 84"
        />
      </svg>
    </slot>

    <span :class="labelClasses">
      <slot>{{ label }}</slot>
    </span>
  </div>
</template>

<style module>
.container {
  display: flex;
  align-items: center;
  gap: var(--ga-size-spacing-02);

  &.vertical {
    flex-direction: column;
    gap: var(--ga-size-spacing-03);
  }
}

.icon {
  &.small {
    width: var(--ga-size-spacing-05);
    height: var(--ga-size-spacing-05);
  }

  &.large {
    width: var(--ga-size-spacing-09);
    height: var(--ga-size-spacing-09);
  }
}

@keyframes spin {
  100% {
    transform: rotate(360deg);
  }
}
.spinner {
  transform-origin: center;
  animation: spin 1.1s cubic-bezier(0.4, 0.2, 0.6, 0.8) infinite;
  will-change: transform;
}

.label {
  min-height: var(--ga-size-spacing-05);
  color: var(--ga-color-text-body);

  font-weight: var(--ga-font-weight-normal);
  font-size: var(--ga-text-sm-font-size);
  line-height: var(--ga-text-sm-line-height);
  font-family: var(--ga-font-family-primary);
  letter-spacing: var(--ga-text-sm-tracking);

  &:empty {
    display: none;
  }
}

.label {
  &.small {
    font-size: var(--ga-text-xs-font-size);
    line-height: var(--ga-text-xs-line-height);
    letter-spacing: var(--ga-text-xs-tracking);
  }

  &.large {
    font-size: var(--ga-text-md-font-size);
    line-height: var(--ga-text-md-line-height);
    letter-spacing: var(--ga-text-md-tracking);
  }
}
</style>
