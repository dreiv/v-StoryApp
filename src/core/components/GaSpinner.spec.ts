import { shallowMount, type VueWrapper } from '@vue/test-utils'
import GaSpinner from './GaSpinner.vue'

vi.mock('../composables/useFormInput')
vi.stubGlobal('window', { matchMedia: vi.fn(() => ({ matches: false })) })

describe('Spinner', () => {
  let wrapper: VueWrapper
  beforeAll(() => {
    wrapper = shallowMount(GaSpinner, { slots: { default: 'mock content' } })
  })

  afterAll(() => {
    vi.unstubAllGlobals()
  })

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should apply the correct container classes based on direction', async () => {
    expect(wrapper.find('.container').classes()).not.toContain('vertical')

    await wrapper.setProps({ direction: 'vertical' })
    expect(wrapper.find('.container').classes()).toContain('vertical')
  })

  it('should apply the correct size classes', async () => {
    await wrapper.setProps({ size: 'sm' })
    expect(wrapper.find('.icon').classes()).toContain('small')
    expect(wrapper.find('.label').classes()).toContain('small')

    await wrapper.setProps({ size: 'lg' })
    expect(wrapper.find('.icon').classes()).toContain('large')
    expect(wrapper.find('.label').classes()).toContain('large')
  })

  it('should pass aria attributes', async () => {
    await wrapper.setProps({
      label: 'Test Label',
      error: true,
      errorMessage: 'Test Error Message',
    })

    const container = wrapper.find('.container')
    expect(container.attributes('aria-label')).toBe('Test Label')
    expect(container.attributes('aria-invalid')).toBe('true')
    expect(container.attributes('aria-errormessage')).toBe('Test Error Message')
  })
})
