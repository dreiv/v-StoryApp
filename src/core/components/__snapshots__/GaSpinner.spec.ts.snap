// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Spinner > should render correctly with default props 1`] = `
"<div class="container"><svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" viewBox="0 0 32 32" stroke-width="6.4" class="icon">
    <defs>
      <circle id="circle" cx="16" cy="16" r="12.8" fill="none"></circle>
    </defs>
    <use href="#circle" stroke="var(--ga-color-secondary)"></use>
    <use href="#circle" class="spinner" stroke="var(--ga-color-primary-70)" stroke-linecap="round" stroke-dasharray="21 84"></use>
  </svg><span class="label"></span></div>"
`;

exports[`Spinner > should render correctly with default slot content 1`] = `
"<div class="container"><svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" viewBox="0 0 32 32" stroke-width="6.4" class="icon">
    <defs>
      <circle id="circle" cx="16" cy="16" r="12.8" fill="none"></circle>
    </defs>
    <use href="#circle" stroke="var(--ga-color-secondary)"></use>
    <use href="#circle" class="spinner" stroke="var(--ga-color-primary-70)" stroke-linecap="round" stroke-dasharray="21 84"></use>
  </svg><span class="label">Custom loading message</span></div>"
`;

exports[`Spinner > should render correctly with icon slot content 1`] = `
"<div class="container">
  <div class="custom-icon">⚙️</div><span class="label"></span>
</div>"
`;

exports[`Spinner > should render correctly with label prop 1`] = `
"<div class="container" aria-label="Loading data"><svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" viewBox="0 0 32 32" stroke-width="6.4" class="icon">
    <defs>
      <circle id="circle" cx="16" cy="16" r="12.8" fill="none"></circle>
    </defs>
    <use href="#circle" stroke="var(--ga-color-secondary)"></use>
    <use href="#circle" class="spinner" stroke="var(--ga-color-primary-70)" stroke-linecap="round" stroke-dasharray="21 84"></use>
  </svg><span class="label">Loading data</span></div>"
`;
